src/hotspot/share/cds/filemap.cpp

```diff
@@ -2678,11 +2681,50 @@ ClassFileStream* FileMapInfo::open_stream_for_jvmti(InstanceKlass* ik, Handle cl
   const char* const file_name = ClassLoader::file_name_for_class_name(class_name,
                                                                       name->utf8_length());
   ClassLoaderData* loader_data = ClassLoaderData::class_loader_data(class_loader());
-  ClassFileStream* cfs = cpe->open_stream_for_loader(THREAD, file_name, loader_data);
+  ClassFileStream* cfs;
+  if (class_loader() != nullptr && !cpe->is_modules_image()) {
+    cfs = get_stream_from_class_loader(class_loader, cpe, file_name, CHECK_NULL);
+  } else {
+    cfs = cpe->open_stream_for_loader(THREAD, file_name, loader_data);
+  }
   assert(cfs != nullptr, "must be able to read the classfile data of shared classes for built-in loaders.");
   log_debug(cds, jvmti)("classfile data for %s [%d: %s] = %d bytes", class_name, path_index,
                         cfs->source(), cfs->length());
   return cfs;
 }
 
+ClassFileStream* FileMapInfo::get_stream_from_class_loader(Handle class_loader,
+                                                           ClassPathEntry* cpe,
+                                                           const char* file_name,
+                                                           TRAPS) {
+  JavaValue result(T_OBJECT);
+  TempNewSymbol class_name_sym = SymbolTable::new_symbol(file_name);
+  Handle ext_class_name = java_lang_String::externalize_classname(class_name_sym, CHECK_NULL);
+
+  // byte[] ClassLoader.getResourceAsByteArray(String name)
+  JavaCalls::call_virtual(&result,
+                          class_loader,
+                          vmClasses::ClassLoader_klass(),
+                          vmSymbols::getResourceAsByteArray_name(),
+                          vmSymbols::getResourceAsByteArray_signature(),
+                          ext_class_name,
+                          CHECK_NULL);
+  assert(result.get_type() == T_OBJECT, "just checking");
+  oop obj = result.get_oop();
+  assert(obj != nullptr, "ClassLoader.getResourceAsByteArray should not return null");
+
+  // The result should be a [B
+  assert(obj->is_typeArray(), "just checking");
+  assert(TypeArrayKlass::cast(obj->klass())->element_type() == T_BYTE, "just checking");
```

Comment:
This seems unnecessary. What kind of errors are you trying to detect with this? The method signature indicates it returns a byte-array.

Reply:
I adapted the asserts from diagnosticCommand.cpp. I'm keeping the first assert. Is it ok?

Reply:
Both asserts are unnecessary.

Reply:
I've removed the assert. Will run some tests before integrating the fix.
Please re-review when you have a chance.